### ISSUE TITLE
Tail-Based Sampling - Scalability Issues

### ISSUE THREAD

**Issue by vikrambe** [open-telemetry/opentelemetry-collector#407]
I like the idea of tail based sampling. I would like to enhance tail based sampling in 2 ways:

Consider a case where we are tail-sampling based on span->http.status_code=500. 
1. I would want to sample 100% of all http.status_code "500"s and 10% of http.status_code "200"s. I would like to create a new sampler that combines string_tag_filter and probabilistic sampler.
2. What would happen when Spans for a trace land-up in different Otel-Collector instances, one instance can sample the trace when it find http.status_code as 500 and the other could ignore the sampling. I was thinking of using group-cache to solve this problem.
Please, share your thoughts on the above.

**Comment by vikrambe** [comment]
Please review fix for #1 above. 
i.e. 
> I would want to sample 100% of all http.status_code "500"s and 10% of http.status_code "200"s. I would like to create a new sampler that combines string_tag_filter and probabilistic sampler.

**Comment by joe-elliott** [comment]
A no external dependencies option might look like:
- Consistent hash ring communicated via a gossip based protocol
- Upon receiving a span a collector hashes the trace id and forwards it to the appropriate collector

I believe this is very roughly what Cassandra does.

During scale up/down traces will be scattered across multiple collectors as the hash ring settles.  I suppose these traces would then not be eligible for tail based sampling and would be subject to some kind of default sampling policy.

**Comment by annanay25** [comment]
I'd like to share the design document for scaling tail based sampling for review - https://docs.google.com/document/d/1HEI8E9VFcv_QoA8k4YLoO8Md7XWiE5vqE3pIhKJozTQ/edit?usp=sharing


/cc @open-telemetry/collector-maintainers @joe-elliott

**Comment by vikrambe** [comment]
@annanay25 I thought on the same lines but making otel implement service discovery seems to be an overkill and this design is not scalable at high load. Every otel collector instance starts distributing spans to respective collector instances. It more challenging when we have to unpack grpc batches and redistribute spans. Failover scenarios and tail sampling decision window adds more edge cases and complexity. We need a consistent and a reliable solution that will work at scale here.

**Comment by vikrambe** [comment]
@annanay25 @joe-elliott Maybe use of a in-memory distributed cache should help here. We can run otel collector in two modes . One mode will fill cache Only and other mode will actually make tail sampling decisions based on filter criteria’s. what do you guys think?

**Comment by vikrambe** [comment]
> I'd like to share the design document for scaling tail based sampling for review - https://docs.google.com/document/d/1HEI8E9VFcv_QoA8k4YLoO8Md7XWiE5vqE3pIhKJozTQ/edit?usp=sharing
> 
> /cc @open-telemetry/collector-maintainers @joe-elliott

@annanay25 Load balancers support ring hash based on header value. OtelCollector when configured with tail sampler can unpack grpc batches and hit the ALBs with span batched by grouping criteria such as TraceID or nodeId etc. And ship spans to respective otel collector instance handling TraceID being processed.

**Comment by joe-elliott** [comment]
@vikrambe   One of our goals with the proposed design was to avoid taking on any external dependencies.  We recognize that at a certain scale tools like load balancers or in memory caches would definitely make things easier.

Perhaps a broader discussion about this particular feature and what kind of external dependencies the otel-collector would be willing to take on is necessary here?

**Comment by vikrambe** [comment]
@joe-elliott I dont think we should add any external dependencies to otel collector. To keep it simple and yet solve this problem. One of the options that i can think of is to enhance otel collector's agent mode to batch spans and add headers that would help in routing the batchs to appropriate otel collector instance. Routing batches based on headers can be done by fairly any Load balancer.

**Comment by jpkrohling** [comment]
>  One of our goals with the proposed design was to avoid taking on any external dependencies.

I can totally understand that, but if there's a high-quality embedded distributed caching library out there, it might be worth using it instead of implementing a new one from scratch. @vikrambe's suggestion of [golang/groupcache](https://github.com/golang/groupcache) might be worth the investigation. 

One advantage I can see when using groupcache instead of the proposed mod N hash is that cache misses would not break the tail-based sampling, as the data will be retrieved from a remote ring member.

**Comment by william-tran** [comment]
If you can use kafka, I think we're pretty close: https://github.com/jaegertracing/jaeger/issues/425#issuecomment-625464012
Use two tiers of otel-collector with kafka in the middle.

**Comment by ehmke** [comment]
I am finding this discussion very interesting. How does this relate to the comment in the last paragraph here: https://github.com/open-telemetry/opentelemetry-collector/blob/master/docs/performance.md

It says that we would need a LB supporting traceId-based routing to scale horizontally. Does this present a practical option when using a certain combination of protocols? For example, I am sending batches of traces via thrift_http. I assume, it would be enough if the batches would only carry traces with the same trace id and if every http request would have the trace-id header of the traces it is transporting.

**Comment by secat** [comment]
@annanay25 I have read your post on [How Grafana Labs enables horizontally scalable tail sampling in the OpenTelemetry Collector](https://grafana.com/blog/2020/06/18/how-grafana-labs-enables-horizontally-scalable-tail-sampling-in-the-opentelemetry-collector/).

Are you planning to contribute your new processor called the [_aggregate processor_](https://github.com/grafana/opentelemetry-collector/tree/scale-tail-sampling/processor/aggregateprocessor) to the [open-telemetry/opentelemetry-collector-contrib](https://github.com/open-telemetry/opentelemetry-collector-contrib) project?

**Comment by morigs** [comment]
I think scalable tail-based sampling is a must-have feature and should be inside main repository.
Also IMHO this feature is required for GA.
So should we wait for PR from @annanay25 or implement something similar?

**Comment by secat** [comment]
Is there any update on how to standardize the tail based sampling horizontal scaling?

**Comment by jpkrohling** [comment]
Not yet, we started a discussion as part of #1724. I do plan on working on a PoC based on my last comments there.

**Comment by annanay25** [comment]
Apologies for being unresponsive here. I somehow managed to miss these notifications.

I would be more than happy to contribute the aggregate processor to the contrib repo. However, some concerns were raised in a previous PR where I attempted the same, so I would like to address them as well this time, and work towards something that the community has a consensus on. @jpkrohling has already done some amazing new work in this area, so it would be best to build upon that.

**Comment by jpkrohling** [comment]
Is this still relevant? Given that we have the load balancer in place now, it should be easy to do a scalable tail based solution now.

**Comment by morigs** [comment]
@jpkrohling Could you please describe in more detail the solution with a load balancer?

**Comment by jpkrohling** [comment]
Sure: you'll need two layers of collectors. The first are the load balancer(s), which would be a simple collector with the load balancer exporter configured: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/master/exporter/loadbalancingexporter

The second would be a cluster of collectors with the [tail-sampling processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/master/processor/tailsamplingprocessor), or the [groupbytrace](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/master/processor/groupbytraceprocessor) processor + your own sampling processor, in case you have your own distribution. 

If that's still too abstract, I can try to come up with a concrete example, but it might take a while.

**Comment by vikrambe** [comment]
@jpkrohling I have a couple of queries.
1. How will it handle async calls/long lived transactions ?
2. Load balancing is one part of this issue. We also need some kind of rule engine to evaluate filter criteria. With AND/OR capabilities to sample the traces. Maybe, do we want to create a separate issue to track it ?

**Comment by vikrambe** [comment]
@jpkrohling Thanks for the loadbalancingexporter. This is helpful...

**Comment by jpkrohling** [comment]
For 1, not sure I understand the problem. Both the tail sampling processor and the groupbytrace will wait for a specific amount of time, under the assumption that all spans for the trace will have arrived by then.

The filter criteria is the second part of the tail sampling processor, known as "policy" there. We started thinking about splitting this into its own processor, but it wouldn't bring anything that doesn't exist already, at least not in the first release.

If none of the existing processors/policies are good for you, you can [build your own](https://medium.com/opentelemetry/extending-the-opentelemetry-collector-with-your-own-components-64c10cf675db) and pack as part of your [custom build](https://medium.com/opentelemetry/building-your-own-opentelemetry-collector-distribution-42337e994b63) of the otelcol.

**Comment by morigs** [comment]
Wow, this is really interesting, didn't know that. I hope this pattern will be well documented in the future.
The only concern is the fact that `loadbalancingexporter` is in the contrib repository. This sounds like not quite official way.
May be it should be moved to the main collector repo then

**Comment by jpkrohling** [comment]
Where would you expect to see this information? I can create some documentation about it. This belongs to the contrib because it's still understood to be a "niche" feature. The more people using this feature and providing feedback, the higher the chances that this will get moved to the core ;-)

**Comment by morigs** [comment]
I don't think this is possible right now, but I would expect to see the balancing setup documentation somewhere [here](https://opentelemetry.io/docs/collector/).
I think load balancing is a must, because scaling is impossible without it.
It would be cool to involve the core team members in this discussion

**Comment by jpkrohling** [comment]
When not doing tail-based sampling, you can do load balancing using the regular techniques. This specialized load balancer is needed only if you need to ensure that the same trace ID is consistently sent to the same backend.

**Comment by morigs** [comment]
I agree. Bu IMHO tail-based sampling is a quite common scenario

**Comment by heruv1m** [comment]
@jpkrohling  is there solution for question from issue topic?

> I would want to sample 100% of all http.status_code "500"s and 10% of http.status_code "200"s

Is seems, that [tail sampling](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/tailsamplingprocessor)  unable to set percent of samples

**Comment by jpkrohling** [comment]
You are right -- policies can't be chained at the moment, though this is something we discussed in the past.

**Comment by heruv1m** [comment]
Can I do that sampling in another way?

**Comment by jpkrohling** [comment]
Not with the current processors. Building one isn't that complicated, though. Here's how the policy parts of the tail-based sampling look like when split from that processor: https://github.com/jpkrohling/opentelemetry-collector-contrib/commit/fbac498f8cb32c95fe15029b4ea29de7aadc5dbe

**Comment by heruv1m** [comment]
I think your processor is what I need.
PR in contrib repo would be nice

**Comment by karthikinamanamelluri** [comment]
@vikrambe 
Could you please refer to the contrib PR here or if there is fork created out of contrib.
